design/service-indicator-lights.md

```diff
@@ -0,0 +1,201 @@
+provide a way to turn those lights on when service is needed for a
+host, and off when service is no longer needed. This document
+describes the new CRD and controller for managing indicator lights.
+
+## Motivation
+
+### Goals
+
+1. Enable operations staff to identify physical devices associated
+   with software-reported faults.
+
+### Non-Goals
+
+1. Actively manage all indicator lights for all known hardware.
+
+## Proposal
+
+### User Stories
+
+#### As an operator, I want to be able to match a software-reported fault to the hardware device that has failed.
+
+### Implementation Details/Notes/Constraints
+
+As with many BMC-related operators, we must take care not to overwhelm
+the management controller with too many requests for state or
+updates. Since a single chassis might have a dozen or more indicator
+lights, polling for their status regularly may place excessive load on
+the management controller.
+
+Unlike some other operations with BMCs, polling or changing the state
+of the indicator light for an internal component may require
+interacting with the I/O controller, which may interrupt regular I/O
+operations. For this reason, Ironic does not regularly poll the status
+of indicator lights and does not cache their status when an outside
+user queries it.
+
+We do not want to disturb production data flow any more than
+necessary, so we want to limit the amount of polling we do related to
+indicator lights from metal3.
+
+Because of these 2 constraints, we do not want to add indicator
+management to the existing BareMetalHost controller or CRD. Instead,
+we will define a new CRD to represent a service indicator light that
+is in use.
+
+Only lights associated with an instance of the new CRD will be managed
+at all, and that management is limited to checking and toggling the
+state of the light when the resource in k8s changes. No active
+management is performed to ensure that state continues to match the
+resource.
+
+### Risks and Mitigations
+
+1. Polling too frequently will cause more than the usual level of
+   issues with the BMC or disk controllers. We mitigate these problems
+   by reducing the number of devices we actually poll, and reducing
+   the polling frequency.
+
+2. Because less polling is performed, the state of the CR is more
+   likely to become out of sync with the hardware if some other
+   component is also manipulating the indicator light. We can mitigate
+   the effect of that by ensuring that the code that turns the light
+   on/off checks the status before trying to change it so it is
+   idempotent.
+
+## Design Details
+
+We will define a new CRD `BareMetalHostServiceIndicator` (referred to
+as "indicator CRD" elsewhere in this document) with the following spec
+and status structures.
+
+```go
+type BareMetalHostServiceIndicatorSpec struct {
+	// HostRef is a reference to the bare metal host
+	HostRef corev1.ObjectReference `json:"hostRef"`
+
+	// Indicator ID
```

Comment:
How do we initially discover these IDs? Currently, ironic uses indicator names as manifested by the Redfish agent.

Reply:
There will be a list on the host object. We will either have the raw list as strings, or if we can map them to devices we can weave them into the existing data structures.

Reply:
I added a few sentences to explain that at the start of this section.

Reply:
Possible instability of the enumeration bothers me....

Imagine we have many drives, Redfish BMC enumerates them somehow. At some point the operator replaces some drives (pulls old and plugs new). What will happen to the initial enumeration? I fear that is vendor-specific.

Arguably, the bullet-proof approach would be to "calibrate" the indicators meaning lighting up each of the discovered LEDs and manually assigning it to corresponding hardware unit...

Reply:
Refreshing the known devices in a host is an orthogonal problem and there's a separate design to discuss that. I would expect the indicator list to be refreshed when the rest of the hardware is.

I don't think it's realistic to expect a user to configure the indicator lights in the way you describe, so we're going to have to do the best we can.

Do we have any idea what an indicator ID will look like for a drive? The examples in the spec at http://specs.openstack.org/openstack/ironic-specs/specs/not-implemented/expose-hardware-indicators.html don't include drives.

Reply:
> Do we have any idea what an indicator ID will look like for a drive?

Ironic take on indicator naming is that it's up to the driver talking to the BMC (AKA *hardware type* in ironic parlance).

When it comes to Redfish, in the schema, the indicator LED resource does not carry any dedicated name. Thus, `redfish` driver in ironic [reuses](https://review.opendev.org/gitweb?p=openstack/ironic.git;f=ironic/drivers/modules/redfish/management.py;hb=refs/changes/40/652740/5#l350) parent resource's UUID for the indicator sitting on it. Well, for as long as it's a single LED per addressable unit.

Reply:
@etingof if you could give me a real example, we could include it in this doc. Do we have hardware that has indicators that we can use to build examples?